src/hotspot/share/opto/escape.cpp

```diff
@@ -4120,6 +4120,22 @@ Node* ConnectionGraph::find_inst_mem(Node *orig_mem, int alias_idx, GrowableArra
       }
       result = result->in(MemNode::Memory);
     }
+    if (!is_instance && result->Opcode() == Op_NarrowMemProj) {
+      // Memory for non known instance can safely skip over a known instance allocation (that memory state doesn't access
+      // the result of an allocation for a known instance).
+      assert(result->as_Proj()->_con == TypeFunc::Memory, "a NarrowMemProj can only be a memory projection");
```

Comment:
Can we verify that already in the `NarrowMemProj`, i.e. its constructor?

Reply:
`TypeFunc::Memory` is now hardcoded in the `NarrowMemProj` constructor.